### PR TITLE
bpo-35505: Skip test_imap4_host_default_value if localhost listens on IMAP port

### PR DESCRIPTION
**Pull request by mcepl**
test_imaplib.TestImaplib.test_imap4_host_default_value fails when run on
the computer with the local IMAP server running. This may happen for
developers, who consequently don't need to be bothered about this test
failing.

https://bugs.python.org/issue35505

### CODE REVIEW

**Review by eamanu — APPROVED** [review]
lgtm. I don't think that is necessary a news for a new test.

**Comment by vstinner** [comment]
Would it be possible to use a different port instead? Using test.support.find_unused_port() for example.

File: Misc/NEWS.d/next/Library/2019-02-12-01-33-08.bpo-35505.N9ba_K.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+test_imaplib.TestImaplib.test_imap4_host_default_value fails when run on the
+computer with the local IMAP server running. This may happen for developers,
+who consequently don't need to be bothered about this test failing when
+working on other CPython patches.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2019-02-12-01-33-08.bpo-35505.N9ba_K.rst]: Hum, I suggest a shorter description: "Fix test_imaplib when a local IMAP server is running."

Comment (mcepl) [reply Misc/NEWS.d/next/Library/2019-02-12-01-33-08.bpo-35505.N9ba_K.rst]: Like this?

**Comment by vstinner** [comment]
I tested manually.

Terminal 1:

```
echo -n|sudo nc -l 143 -v
```

Terminal 2:

```
./python -m test -u all test_imaplib -m test_imap4_host_default_value -v
```

Without this change, Terminal 1 shows an incoming connection ("Ncat: Connection from ::1:35778.") and the test fails with imaplib.IMAP4.abort: socket error: EOF.

With the change: no incoming connection and the test pass as expected.

File: Lib/test/test_imaplib.py [hunk 2]
```diff
@@ -70,7 +70,22 @@ def test_that_Time2Internaldate_returns_a_result(self):
         for t in self.timevalues():
             imaplib.Time2Internaldate(t)
 
+    @staticmethod
+    def _check_port(address='localhost', port=imaplib.IMAP4_PORT):
+        import socket
```
Comment (vstinner) [inline Lib/test/test_imaplib.py]: I suggest to move the import at top level.

File: Lib/test/test_imaplib.py [hunk 3]
```diff
@@ -70,7 +70,22 @@ def test_that_Time2Internaldate_returns_a_result(self):
         for t in self.timevalues():
             imaplib.Time2Internaldate(t)
 
+    @staticmethod
+    def _check_port(address='localhost', port=imaplib.IMAP4_PORT):
```
Comment (vstinner) [inline Lib/test/test_imaplib.py]: The function is only used once. You can just inline the code directly into test_imap4_host_default_value().

It seems like IMAP4 uses '' (empty string) as the default host, not "localhost" which might be a different host depending on the config.

File: Lib/test/test_imaplib.py [hunk 4]
```diff
@@ -70,7 +70,22 @@ def test_that_Time2Internaldate_returns_a_result(self):
...
+        with socket.socket() as s:
+            try:
+                s.connect((address, port))
+                return True
+            except socket.error:
+                return False
+
     def test_imap4_host_default_value(self):
+        if self._check_port():
+            raise unittest.SkipTest(
```
Comment (vstinner) [inline Lib/test/test_imaplib.py]: nitpick: I prefer self.skipTest("...") ;-)

File: Misc/NEWS.d/next/Library/2019-02-12-01-33-08.bpo-35505.N9ba_K.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Make test_imap4_host_default_value independent on whether the
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2019-02-12-01-33-08.bpo-35505.N9ba_K.rst]: Oh, I missed the filename: please move this NEWS entry to the Tests/ category.

File: Lib/test/test_imaplib.py (hunk 2 above)
Comment (mcepl) [reply Lib/test/test_imaplib.py]: What about no import whatsoever? ;)

**Review by vstinner — APPROVED** [review]
LGTM.

File: Lib/test/test_imaplib.py [hunk 6]
```diff
@@ -71,6 +71,15 @@ def test_that_Time2Internaldate_returns_a_result(self):
             imaplib.Time2Internaldate(t)
 
     def test_imap4_host_default_value(self):
+        # Check whether the IMAP4_PORT is truly unavailable.
+        with imaplib.socket.socket() as s:
```
Comment (vstinner) [inline Lib/test/test_imaplib.py]: Hum wait, no, you cannot get socket from imaplib. You have to import is explicitly. imaplib.socket can disappear anytime and it's not part of it's public API. The test must not rely on implementation details.

Comment (mcepl) [reply Lib/test/test_imaplib.py]: oh well, write hundred times "You shall not be clever."

**Comment by vstinner** [comment]
@mcepl: The patch doesn't apply cleanup on the 2.7 branch. Can you try to write at patch for 2.7?

**Comment by mcepl** [comment]
@vstinner There is nothing to port to … 2.7 just doesn't have this test case at all.

**Comment by vstinner** [comment]
> @vstinner There is nothing to port to … 2.7 just doesn't have this test case at all.

Aha :-D Good. I close the issue in that case ;-)

**Pull request merged**